pkg/images/set_test.go

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2020-Present Pivotal Software, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package images_test
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/pivotal/image-relocation/pkg/image"
+	"github.com/pivotal/image-relocation/pkg/images"
+)
+
+var _ = Describe("FlattenRepoPath", func() {
+	Describe("New", func() {
+		var (
+			ii  []string
+			s   images.Set
+			err error
+		)
+
+		JustBeforeEach(func() {
+			s, err = images.New(ii...)
+		})
+
+		Context("when the input slice is empty", func() {
+			BeforeEach(func() {
+				ii = []string{}
+			})
+
+			It("should construct an empty set", func() {
+				Expect(err).NotTo(HaveOccurred())
+				Expect(s).To(Equal(images.Empty))
+			})
+		})
+
+		Context("when the input slice is non-empty", func() {
+			BeforeEach(func() {
+				ii = []string{"a", "b"}
+			})
+
+			It("should construct the correct set", func() {
+				Expect(err).NotTo(HaveOccurred())
+
+				v, err := images.New("b", "a")
+				Expect(err).NotTo(HaveOccurred())
+
+				Expect(s).To(Equal(v))
+			})
+		})
+
+		Context("when the input slice has an invalid image reference", func() {
+			BeforeEach(func() {
+				ii = []string{"::"}
+			})
+
+			It("should return a suitable error", func() {
+				Expect(err).To(MatchError(`invalid image reference: "::"`))
+			})
+		})
+	})
+
+	Describe("Union", func() {
+		It("should produce the correct result", func() {
+			s, err := images.New("a")
+			Expect(err).NotTo(HaveOccurred())
+
+			u, err := images.New("b")
+			Expect(err).NotTo(HaveOccurred())
+
+			v, err := images.New("a", "b")
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(s.Union(u)).To(Equal(v))
+		})
+	})
+
+	Describe("Slice", func() {
+		It("should produce the correct result", func() {
+			s, err := images.New("a", "b")
+			Expect(err).NotTo(HaveOccurred())
+
+			a, err := image.NewName("a")
+			Expect(err).NotTo(HaveOccurred())
+
+			b, err := image.NewName("b")
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(s.Slice()).To(ConsistOf(a, b))
+		})
+	})
+
+	Describe("Strings", func() {
+		It("should produce the correct result", func() {
+			s, err := images.New("b", "a")
+			Expect(err).NotTo(HaveOccurred())
+
+			Expect(s.Strings()).To(Equal([]string{"docker.io/library/a", "docker.io/library/b"}))
```

Comment:
I don't completely understand where these `"docker.io/library"` prefixes come from

Reply:
Image references like "a" are implicitly on docker hub (hence "docker.io") and belong to the user "library". This behaviour is seen in the image name tests. But you'r right that this is a little distracting here, we I'll switch to using image references like "example.com/u/a" instead of "a".

Reply:
ok, the "library" lost me, but it makes sense after your explanation, thanks